crates/extensions/c8y_mapper_ext/src/converter.rs

```diff
@@ -682,12 +742,41 @@ impl Converter for CumulocityConverter {
     fn get_mapper_config(&self) -> &MapperConfig {
         &self.mapper_config
     }
+
     async fn try_convert(&mut self, message: &Message) -> Result<Vec<Message>, ConversionError> {
-        match &message.topic {
-            topic if topic.name.starts_with("tedge/measurements") => {
-                self.size_threshold.validate(message)?;
-                self.try_convert_measurement(message)
+        let mut registration_messages = vec![];
+
+        if let Some(entity_id) = entity_store::entity_mqtt_id(&message.topic) {
+            if is_entity_register_message(message) {
+                if let Err(e) = self
+                    .entity_store
+                    .update(message.clone().try_into().unwrap())
+                {
+                    error!("Could not update device registration: {e}");
+                }
+            } else {
+                // if device is unregistered register using auto-registration
+                if self.entity_store.get(entity_id).is_none() {
+                    registration_messages = match self.auto_register_entity(entity_id) {
+                        Ok(register_messages) => register_messages,
+                        Err(e) => {
+                            error!("Could not update device registration: {e}");
+                            vec![]
+                        }
+                    };
+
+                    let entity = self.entity_store.get(entity_id).unwrap();
+                    if let Some(message) = external_device_registration_message(entity) {
+                        self.children
+                            .insert(entity.external_id.to_string(), Operations::default());
+                        registration_messages.push(message);
+                    }
+                }
             }
+        }
+
+        let mut messages = match &message.topic {
+            topic if EntityTopic::is_measurement(topic) => self.try_convert_measurement(message),
```

Comment:
Here, I would prefer to have a fully different `match` expression.

Instead of having methods working on raw topics and messages,  my goal is to do all the parsing inside the `tedge-api` crate with some function taking an MqttMessage and returning an error or a plain struct with the message source (e.g. `device/main//`) the channel (e.g. Measurement { r#type })  and the decoded message (e.g the measurement). The converter main method would be then a breeze focused on the translation into the format expected by the cloud and no more a mix of parsing / validation / translation.